src/FavIconManager.vala

```diff
@@ -0,0 +1,321 @@
+//	This file is part of FeedReader.
+//
+//	FeedReader is free software: you can redistribute it and/or modify
+//	it under the terms of the GNU General Public License as published by
+//	the Free Software Foundation, either version 3 of the License, or
+//	(at your option) any later version.
+//
+//	FeedReader is distributed in the hope that it will be useful,
+//	but WITHOUT ANY WARRANTY; without even the implied warranty of
+//	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//	GNU General Public License for more details.
+//
+//	You should have received a copy of the GNU General Public License
+//	along with FeedReader.  If not, see <http://www.gnu.org/licenses/>.
+
+public class FeedReader.FavIconManager : GLib.Object {
+
+	private Gee.HashMap<string, Gdk.Pixbuf> m_map;
+	private static FavIconManager? m_cache = null;
+
+	public signal void ReloadFavIcon(Feed feed);
+
+	public static FavIconManager get_default()
+	{
+		if(m_cache == null)
+			m_cache = new FavIconManager();
+
+		return m_cache;
+	}
+
+	private FavIconManager()
+	{
+		m_map = new Gee.HashMap<string, Gdk.Pixbuf>();
+	}
+
+	private async void load(Feed feed, bool firstTry = true)
+	{
+		var fileName = GLib.Base64.encode(feed.getFeedID().data) + ".ico";
+		try
+		{
+			var file = File.new_for_path(GLib.Environment.get_user_data_dir() + "/feedreader/data/feed_icons/" + fileName);
+			var stream = yield file.read_async();
+			var pixbuf = yield new Gdk.Pixbuf.from_stream_async(stream);
+			stream.close();
+			if(pixbuf.get_height() <= 1 && pixbuf.get_width() <= 1)
+			{
+				Logger.warning(@"FavIconManager: $fileName is too small");
+				return;
+			}
+
+			pixbuf = pixbuf.scale_simple(24, 24, Gdk.InterpType.BILINEAR);
+			m_map.set(feed.getFeedID(), pixbuf);
+			ReloadFavIcon(feed);
+		}
+		catch (IOError.NOT_FOUND e)
+		{
+			//Logger.debug(@"FavIconManager: Icon $fileName does not exist");
+
+			if(!firstTry)
+				return;
+
+			bool success = yield downloadFavIcon(feed);
+			if(success)
+				yield load(feed, false);
```

Comment:
I think that if an icon doesn't exist, this will cause us to try downloading it every time we display one (since `firstTry` is an argument to the function, not a member variable).

Having a `m_icons_loading` list (or set) would let us keep track of which ones we already have a download running for, but we'd need to somehow link that to the currently executing async function. This seems like obvious case where we should use Promises but I don't think they exist in Vala, and I'm not sure what the best alternative would be. Maybe exclusively using the signals and not have the `getIcon` function?

Reply:
Wouldn't it just load the existing metadata in that case and see that a download was already attemted recently?

Reply:
Yes that! We can have `private Gee.HashMap<string, Gee.Future<Gdk.Pixbuf>> m_map;` then this should work:

```
public async Gdk.Pixbuf? getIcon(Feed feed)
{
    var feed_id = feed.getFeedID();
    var future = m_map.get(feed_id);

    if(future == null)
    {
        var promise = new Gee.Promise<Gdk.Pixbuf?>();
        future = promise.future;
        m_map.set(feed_id, future);
        load.begin(feed, (obj, res) => {
            try
            {
                promise.set_value(load.end(res));
            }
            catch(Error e)
            {
                promise.set_error(e);
            }
        });
    }

    // Assuming Gee.Future works properly, we should be able
    // to call these multiple times, and after the first time, it
    // should return instantly
    yield future.wait_async();
    return future.get_value();
}
```

Reply:
Would you like to commit these changes? Otherwise I'll try to take a look at it tonight.

Reply:
Sure I'll try

Reply:
Actually it's too complicated for me to do right now. I'll try to take a look tonight if you haven't already.

Reply:
Almost have this working..